File: DigiMox/app/src/main/java/com/digimox/fragments/DMMainCategoryFragment.java
package com.digimox.fragments;

import android.content.res.Configuration;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.RelativeLayout;

import com.digimox.R;
import com.digimox.activity.DMHomeActivity;
import com.digimox.adapters.DMCategoryAdapter;
import com.digimox.api.DMApiManager;
import com.digimox.app.DMAppConstants;
import com.digimox.models.response.DMMainCategory;
import com.digimox.models.response.DMUserDetails;
import com.digimox.utils.DMUtils;
import com.google.gson.Gson;
import com.loopj.android.http.JsonHttpResponseHandler;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

import cz.msebera.android.httpclient.Header;

/**
 * Created by Deepesh on 26-Nov-15.
 */
public class DMMainCategoryFragment extends DMBaseFragment {
    private RecyclerView mainCategoryList;
    private View view;
    public DMUserDetails dmUserDetails;
    private SwipeRefreshLayout swipeRefreshLayout;
    private RelativeLayout pullLayout;
    private Handler pullHandler = new Handler();


    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_main_category, container, false);
        initViews();
        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        getMainCategory();
    }

    private void getMainCategory() {
        if (DMUtils.isOnline()) {
            requestDidStart();
            String url = DMApiManager.METHOD_MAIN_CATEGORY + "gid=0&lid=" + DMUtils.getLanguageId(getActivity()) + "&uid=" + DMUtils.getUserId(getActivity());
            DMApiManager dmApiManager = new DMApiManager(getActivity());
            dmApiManager.get(url, new JsonHttpResponseHandler() {
                @Override
                public void onSuccess(int statusCode, Header[] headers, JSONArray response) {
                    super.onSuccess(statusCode, headers, response);
                    requestDidFinish();
                    try {
                        ArrayList<DMMainCategory> dmMainCategories = new ArrayList<DMMainCategory>();
                        Gson gson = new Gson();
                        for (int i = 0; i < response.length() - 1; i++) {
                            JSONObject responseJSon = response.getJSONObject(i);
                            DMMainCategory dmMainCategory = gson.fromJson(responseJSon.toString(), DMMainCategory.class);
                            dmMainCategories.add(dmMainCategory);
                        }
                        initViews(dmMainCategories);

                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                }
            });
        } else {
            if (isAdded())
                showToast(getResources().getString(R.string.api_error_no_network));
        }

    }

    private void initViews() {
        mainCategoryList = (RecyclerView) view.findViewById(R.id.main_category_list);
        swipeRefreshLayout = (SwipeRefreshLayout) view.findViewById(R.id.swipeRefreshLayout);
        pullLayout = (RelativeLayout) view.findViewById(R.id.pull_down_layout);
        mainCategoryList.setHasFixedSize(true);
        Animation animation = AnimationUtils.loadAnimation(getActivity(), R.anim.slide_in_up);
        animation.setDuration(1000);
        pullLayout.setVisibility(View.VISIBLE);
        pullLayout.setAnimation(animation);
        pullLayout.animate();
        animation.start();
        animation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
                pullHandler.postDelayed(pullRunnable, 2000);
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });
    }

    private void initViews(ArrayList<DMMainCategory> dmMainCategories) {
        mainCategoryList.setAdapter(null);
        checkOrientation();
        DMCategoryAdapter dmCategoryAdapter = new DMCategoryAdapter(getActivity(), this, dmMainCategories);
        mainCategoryList.setAdapter(dmCategoryAdapter);
        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                getMainCategory();
            }
        });
        swipeRefreshLayout.setRefreshing(false);


    }

    //Here's a runnable/handler combo
    private Runnable pullRunnable = new Runnable() {
        @Override
        public void run() {
            if (null != getActivity()) {
                pullLayout.clearAnimation();
                Animation animation = AnimationUtils.loadAnimation(getActivity(), R.anim.slide_out_up);
                animation.setDuration(1000);
                pullLayout.setAnimation(animation);
                pullLayout.animate();
                animation.start();
                animation.setAnimationListener(new Animation.AnimationListener() {
                    @Override
                    public void onAnimationStart(Animation animation) {

                    }

                    @Override
                    public void onAnimationEnd(Animation animation) {
                        pullLayout.setVisibility(View.GONE);
                    }

                    @Override
                    public void onAnimationRepeat(Animation animation) {

                    }

                });
            }
        }
    };

    private void checkOrientation() {
        int orientation = DMUtils.getScreenOrientation(getActivity());
        if (DMAppConstants.PORTRAIT == orientation) {
            mainCategoryList.setLayoutManager(new GridLayoutManager(getActivity(), 2));
        } else {
            mainCategoryList.setLayoutManager(new GridLayoutManager(getActivity(), 4));
        }
    }

    public void clickMainCategory(ArrayList<DMMainCategory> dmMainCategories, DMMainCategory dmMainCategory, int position) {
        if (0 < Integer.parseInt(dmMainCategory.getHasChild())) {
            showMainCategory(dmMainCategory);
        } else {
            showSubCategoryMainFragment(dmMainCategories, dmMainCategory, position);
        }
    }

    private void showMainCategory(DMMainCategory dmMainCategory) {
        final android.support.v4.app.FragmentManager fm = getActivity().getSupportFragmentManager();
        if (null != fm) {
            DMMainCategorySubFragment dmMainCategorySubFragment = new DMMainCategorySubFragment();
            dmMainCategorySubFragment.setDmMainCategory(dmMainCategory);
            final android.support.v4.app.FragmentTransaction ft = fm.beginTransaction();
            ft.add(R.id.fragment_container, dmMainCategorySubFragment, "mainCategorySubFragment")
                    .addToBackStack("mainCategorySubFragment")
                    .commit();
        }
    }

    private void showSubCategoryMainFragment(ArrayList<DMMainCategory> dmMainCategories, DMMainCategory dmMainCategory, int position) {
        final android.support.v4.app.FragmentManager fm = getActivity().getSupportFragmentManager();
        if (null != fm) {
            DMSubCategoryParentFragment dmSubCategoryParentFragment = new DMSubCategoryParentFragment();
            dmSubCategoryParentFragment.setDmMainCategory(dmMainCategory);
            dmSubCategoryParentFragment.setDmMainCategories(dmMainCategories);
            dmSubCategoryParentFragment.setPosition(position);
            final android.support.v4.app.FragmentTransaction ft = fm.beginTransaction();
            ft.setCustomAnimations(R.anim.slide_in_right, R.anim.slide_out_left, R.anim.slide_in_right, R.anim.slide_out_left)
                    .add(R.id.fragment_container, dmSubCategoryParentFragment, "subCategoryMainFragment")
                    .addToBackStack("subCategoryMainFragment")
                    .commit();
            ((DMHomeActivity) getActivity()).showBackArrow();
        }
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        int orientation = newConfig.orientation;
        switch (orientation) {
            case Configuration.ORIENTATION_LANDSCAPE:
                mainCategoryList.setLayoutManager(new GridLayoutManager(getActivity(), 4));
                break;
            case Configuration.ORIENTATION_PORTRAIT:
                mainCategoryList.setLayoutManager(new GridLayoutManager(getActivity(), 2));
                break;
        }
    }
}

File: DigiMox/app/src/main/java/com/digimox/fragments/DMMainCategoryFragment.kt
package com.digimox.fragments

import android.content.res.Configuration
import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.RelativeLayout
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.digimox.R
import com.digimox.activity.DMHomeActivity
import com.digimox.adapters.DMCategoryAdapter
import com.digimox.api.DMApiManager
import com.digimox.app.DMAppConstants
import com.digimox.models.response.DMMainCategory
import com.digimox.models.response.DMUserDetails
import com.digimox.utils.DMUtils
import com.google.gson.Gson
import com.loopj.android.http.JsonHttpResponseHandler
import cz.msebera.android.httpclient.Header
import org.json.JSONArray
import org.json.JSONException
import java.util.*

/**
 * Created by Deepesh on 26-Nov-15.
 */
class DMMainCategoryFragment : DMBaseFragment() {
    private var mainCategoryList: RecyclerView? = null
    private var itemView: View? = null
    var dmUserDetails: DMUserDetails? = null
    private var swipeRefreshLayout: SwipeRefreshLayout? = null
    private var pullLayout: RelativeLayout? = null
    private val pullHandler = Handler()

    //Here's a runnable/handler combo
    private val pullRunnable = Runnable {
        if (null != activity) {
            pullLayout!!.clearAnimation()
            val animation = AnimationUtils.loadAnimation(activity, R.anim.slide_out_up)
            animation.duration = 1000
            pullLayout!!.animation = animation
            pullLayout!!.animate()
            animation.start()
            animation.setAnimationListener(object : Animation.AnimationListener {
                override fun onAnimationStart(animation: Animation) {

                }

                override fun onAnimationEnd(animation: Animation) {
                    pullLayout!!.visibility = View.GONE
                }

                override fun onAnimationRepeat(animation: Animation) {

                }

            })
        }
    }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        itemView = inflater.inflate(R.layout.fragment_main_category, container, false)
        initViews()
        return itemView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        getMainCategory()
    }

    private fun getMainCategory() {
        if (DMUtils.isOnline()) {
            requestDidStart()
            val url = DMApiManager.METHOD_MAIN_CATEGORY + "gid=0&lid=" + DMUtils.getLanguageId(activity!!) + "&uid=" + DMUtils.getUserId(activity!!)
            val dmApiManager = DMApiManager(activity!!)
            dmApiManager[url, object : JsonHttpResponseHandler() {
                override fun onSuccess(statusCode: Int, headers: Array<Header>?, response: JSONArray?) {
                    super.onSuccess(statusCode, headers, response)
                    requestDidFinish()
                    try {
                        val dmMainCategories = ArrayList<DMMainCategory>()
                        val gson = Gson()
                        for (i in 0 until response!!.length() - 1) {
                            val responseJSon = response.getJSONObject(i)
                            val dmMainCategory = gson.fromJson(responseJSon.toString(), DMMainCategory::class.java)
                            dmMainCategories.add(dmMainCategory)
                        }
                        initViews(dmMainCategories)

                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }

                }
            }]
        } else {
            if (isAdded)
                showToast(resources.getString(R.string.api_error_no_network))
        }

    }

    private fun initViews() {
        mainCategoryList = itemView!!.findViewById<View>(R.id.main_category_list) as RecyclerView
        swipeRefreshLayout = itemView!!.findViewById<View>(R.id.swipeRefreshLayout) as SwipeRefreshLayout
        pullLayout = itemView!!.findViewById<View>(R.id.pull_down_layout) as RelativeLayout
        mainCategoryList!!.setHasFixedSize(true)
        val animation = AnimationUtils.loadAnimation(activity, R.anim.slide_in_up)
        animation.duration = 1000
        pullLayout!!.visibility = View.VISIBLE
        pullLayout!!.animation = animation
        pullLayout!!.animate()
        animation.start()
        animation.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {

            }

            override fun onAnimationEnd(animation: Animation) {
                pullHandler.postDelayed(pullRunnable, 2000)
            }

            override fun onAnimationRepeat(animation: Animation) {

            }
        })
    }

    private fun initViews(dmMainCategories: ArrayList<DMMainCategory>) {
        mainCategoryList!!.adapter = null
        checkOrientation()
        val dmCategoryAdapter = DMCategoryAdapter(activity!!, this, dmMainCategories)
        mainCategoryList!!.adapter = dmCategoryAdapter
        swipeRefreshLayout!!.setOnRefreshListener { getMainCategory() }
        swipeRefreshLayout!!.isRefreshing = false


    }

    private fun checkOrientation() {
        val orientation = DMUtils.getScreenOrientation(activity)
        if (DMAppConstants.PORTRAIT == orientation) {
            mainCategoryList!!.layoutManager = GridLayoutManager(activity, 2)
        } else {
            mainCategoryList!!.layoutManager = GridLayoutManager(activity, 4)
        }
    }

    fun clickMainCategory(dmMainCategories: ArrayList<DMMainCategory>, dmMainCategory: DMMainCategory, position: Int) {
        if (0 < Integer.parseInt(dmMainCategory.hasChild)) {
            showMainCategory(dmMainCategory)
        } else {
            showSubCategoryMainFragment(dmMainCategories, dmMainCategory, position)
        }
    }

    private fun showMainCategory(dmMainCategory: DMMainCategory) {
        val fm = activity!!.supportFragmentManager
        if (null != fm) {
            val dmMainCategorySubFragment = DMMainCategorySubFragment()
            dmMainCategorySubFragment.setMainCategory(dmMainCategory)
            val ft = fm.beginTransaction()
            ft.add(R.id.fragment_container, dmMainCategorySubFragment, "mainCategorySubFragment")
                    .addToBackStack("mainCategorySubFragment")
                    .commit()
        }
    }

    private fun showSubCategoryMainFragment(dmMainCategories: ArrayList<DMMainCategory>, dmMainCategory: DMMainCategory, position: Int) {
        val fm = activity!!.supportFragmentManager
        if (null != fm) {
            val dmSubCategoryParentFragment = DMSubCategoryParentFragment()
            dmSubCategoryParentFragment.setDmMainCategory(dmMainCategory)
            dmSubCategoryParentFragment.setDmMainCategories(dmMainCategories)
            dmSubCategoryParentFragment.setPosition(position)
            val ft = fm.beginTransaction()
            ft.setCustomAnimations(R.anim.slide_in_right, R.anim.slide_out_left, R.anim.slide_in_right, R.anim.slide_out_left)
                    .add(R.id.fragment_container, dmSubCategoryParentFragment, "subCategoryMainFragment")
                    .addToBackStack("subCategoryMainFragment")
                    .commit()
            (activity as DMHomeActivity).showBackArrow()
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        val orientation = newConfig.orientation
        when (orientation) {
            Configuration.ORIENTATION_LANDSCAPE -> mainCategoryList!!.layoutManager = GridLayoutManager(activity, 4)
            Configuration.ORIENTATION_PORTRAIT -> mainCategoryList!!.layoutManager = GridLayoutManager(activity, 2)
        }
    }
}
